In [2]:
import pandas as pd

# Assuming the file is in your Google Drive, in the root 'MyDrive' folder.
df = pd.read_csv('/content/imdb_combined.csv')

# Display the first few rows of the DataFrame to verify
print(df.head())

                                              review  sentiment original_split
0  If you enjoy the original SNL cast and shows t...          0          train
1  A very tired looking Burt Reynolds plays a mer...          0          train
2  First of all, if you'r a fan of the comic, wel...          0          train
3  It was all over with the slashers around 88 so...          0          train
4  I really tried to like this movie. It deals wi...          0          train


In [3]:
# Remove the 'original_split' column (the third column)
df = df.drop(columns=['original_split'], errors = 'ignore')

# Display the first few rows of the updated DataFrame to verify
print(df.head())

                                              review  sentiment
0  If you enjoy the original SNL cast and shows t...          0
1  A very tired looking Burt Reynolds plays a mer...          0
2  First of all, if you'r a fan of the comic, wel...          0
3  It was all over with the slashers around 88 so...          0
4  I really tried to like this movie. It deals wi...          0


### Data Preprocessing Steps

1.  **Check for Missing Values**: Identify and handle any null entries.
2.  **Text Cleaning**: Clean the 'review' column by converting text to lowercase and removing punctuation. This is a basic step to prepare text data for further analysis.

In [4]:
print('Checking for missing values:')
print(df.isnull().sum())

# As there are no missing values based on the output, we can proceed with text cleaning.


Checking for missing values:
review       0
sentiment    0
dtype: int64


In [5]:
import re

def clean_text(text):
    text = text.lower()  # Convert to lowercase
    text = re.sub(r'[^a-z0-9\s]', '', text)  # Remove punctuation and special characters, keep alphanumeric and spaces
    return text

print('\nApplying text cleaning to the \'review\' column...')
df['review'] = df['review'].apply(clean_text)

print('\nPreprocessing complete. Displaying info and first few rows of the cleaned data:')
df.info()
print(df.head())


Applying text cleaning to the 'review' column...

Preprocessing complete. Displaying info and first few rows of the cleaned data:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review     50000 non-null  object
 1   sentiment  50000 non-null  int64 
dtypes: int64(1), object(1)
memory usage: 781.4+ KB
                                              review  sentiment
0  if you enjoy the original snl cast and shows t...          0
1  a very tired looking burt reynolds plays a mer...          0
2  first of all if your a fan of the comic well y...          0
3  it was all over with the slashers around 88 so...          0
4  i really tried to like this movie it deals wit...          0


### Text Tokenization

Tokenization is the process of breaking text into smaller units called tokens (e.g., words, phrases, symbols). This is a fundamental step in most NLP tasks.

In [6]:
import nltk
nltk.download('punkt') # Download the necessary tokenizer data
nltk.download('punkt_tab') # Download 'punkt_tab' as suggested by the error

print('\nTokenizing the \'review\' column...')
df['review_tokens'] = df['review'].apply(nltk.word_tokenize)

print('\nTokenization complete. Displaying the first few rows with tokens:')
print(df[['review', 'review_tokens']].head())

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
[nltk_data] Downloading package punkt_tab to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt_tab.zip.



Tokenizing the 'review' column...

Tokenization complete. Displaying the first few rows with tokens:
                                              review  \
0  if you enjoy the original snl cast and shows t...   
1  a very tired looking burt reynolds plays a mer...   
2  first of all if your a fan of the comic well y...   
3  it was all over with the slashers around 88 so...   
4  i really tried to like this movie it deals wit...   

                                       review_tokens  
0  [if, you, enjoy, the, original, snl, cast, and...  
1  [a, very, tired, looking, burt, reynolds, play...  
2  [first, of, all, if, your, a, fan, of, the, co...  
3  [it, was, all, over, with, the, slashers, arou...  
4  [i, really, tried, to, like, this, movie, it, ...  


### Stop Word Removal

Stop words are words that are filtered out before or after processing of natural language data. Removing them helps in focusing on more important words for analysis and reducing dimensionality.

In [7]:
from nltk.corpus import stopwords
nltk.download('stopwords') # Download the necessary stop words data

stop_words = set(stopwords.words('english'))

def remove_stopwords(tokens):
    return [word for word in tokens if word not in stop_words]

print('\nRemoving stop words from the tokenized reviews...')
df['review_tokens_cleaned'] = df['review_tokens'].apply(remove_stopwords)

print('\nStop word removal complete. Displaying the first few rows with cleaned tokens:')
print(df[['review', 'review_tokens', 'review_tokens_cleaned']].head())

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.



Removing stop words from the tokenized reviews...

Stop word removal complete. Displaying the first few rows with cleaned tokens:
                                              review  \
0  if you enjoy the original snl cast and shows t...   
1  a very tired looking burt reynolds plays a mer...   
2  first of all if your a fan of the comic well y...   
3  it was all over with the slashers around 88 so...   
4  i really tried to like this movie it deals wit...   

                                       review_tokens  \
0  [if, you, enjoy, the, original, snl, cast, and...   
1  [a, very, tired, looking, burt, reynolds, play...   
2  [first, of, all, if, your, a, fan, of, the, co...   
3  [it, was, all, over, with, the, slashers, arou...   
4  [i, really, tried, to, like, this, movie, it, ...   

                               review_tokens_cleaned  
0  [enjoy, original, snl, cast, shows, avoid, mov...  
1  [tired, looking, burt, reynolds, plays, mercen...  
2  [first, fan, comic, well, y

### Data Quality Report

This section provides a summary of the dataset's quality, including sentiment distribution, missing values, duplicates, and uniqueness of reviews.

In [8]:
print('--- Sentiment Distribution ---')
print(df['sentiment'].value_counts())

print('\n--- Missing Values ---')
print(df.isnull().sum())

print('\n--- Duplicate Reviews ---')
duplicate_reviews = df.duplicated(subset=['review']).sum()
print(f'Number of duplicate reviews: {duplicate_reviews}')

print('\n--- Unique Reviews ---')
unique_reviews_count = df['review'].nunique()
print(f'Number of unique reviews: {unique_reviews_count}')

print('\n--- DataFrame Information ---')
df.info()

--- Sentiment Distribution ---
sentiment
0    25000
1    25000
Name: count, dtype: int64

--- Missing Values ---
review                   0
sentiment                0
review_tokens            0
review_tokens_cleaned    0
dtype: int64

--- Duplicate Reviews ---
Number of duplicate reviews: 419

--- Unique Reviews ---
Number of unique reviews: 49581

--- DataFrame Information ---
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 4 columns):
 #   Column                 Non-Null Count  Dtype 
---  ------                 --------------  ----- 
 0   review                 50000 non-null  object
 1   sentiment              50000 non-null  int64 
 2   review_tokens          50000 non-null  object
 3   review_tokens_cleaned  50000 non-null  object
dtypes: int64(1), object(3)
memory usage: 1.5+ MB


### Removing Duplicate Reviews

Based on the quality report, there are `419` duplicate reviews. These will now be removed to ensure that each review entry is unique.

In [9]:
# Remove duplicate reviews based on the 'review' column
df.drop_duplicates(subset=['review'], inplace=True)

print('Duplicate reviews removed. Verifying...')

# Re-check for duplicate reviews
duplicate_reviews_after_removal = df.duplicated(subset=['review']).sum()
print(f'Number of duplicate reviews after removal: {duplicate_reviews_after_removal}')

# Display the new shape of the DataFrame
print(f'New DataFrame shape: {df.shape}')

Duplicate reviews removed. Verifying...
Number of duplicate reviews after removal: 0
New DataFrame shape: (49581, 4)


### Stop Word Removal

Stop words are words that are filtered out before or after processing of natural language data. Removing them helps in focusing on more important words for analysis and reducing dimensionality.

In [10]:
from nltk.corpus import stopwords
nltk.download('stopwords') # Download the necessary stop words data

stop_words = set(stopwords.words('english'))

def remove_stopwords(tokens):
    return [word for word in tokens if word not in stop_words]

print('\nRemoving stop words from the tokenized reviews...')
df['review_tokens_cleaned'] = df['review_tokens'].apply(remove_stopwords)

print('\nStop word removal complete. Displaying the first few rows with cleaned tokens:')
print(df[['review', 'review_tokens', 'review_tokens_cleaned']].head())

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!



Removing stop words from the tokenized reviews...

Stop word removal complete. Displaying the first few rows with cleaned tokens:
                                              review  \
0  if you enjoy the original snl cast and shows t...   
1  a very tired looking burt reynolds plays a mer...   
2  first of all if your a fan of the comic well y...   
3  it was all over with the slashers around 88 so...   
4  i really tried to like this movie it deals wit...   

                                       review_tokens  \
0  [if, you, enjoy, the, original, snl, cast, and...   
1  [a, very, tired, looking, burt, reynolds, play...   
2  [first, of, all, if, your, a, fan, of, the, co...   
3  [it, was, all, over, with, the, slashers, arou...   
4  [i, really, tried, to, like, this, movie, it, ...   

                               review_tokens_cleaned  
0  [enjoy, original, snl, cast, shows, avoid, mov...  
1  [tired, looking, burt, reynolds, plays, mercen...  
2  [first, fan, comic, well, y

In [11]:
from sklearn.model_selection import train_test_split

# Define features (X) and target (y)
X = df['review_tokens_cleaned'] # Using the cleaned, tokenized reviews
y = df['sentiment']

# Split the data into training and testing sets
# Stratify by 'sentiment' to ensure an even distribution of classes in both sets
# random_state for reproducibility
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, stratify=y, random_state=42)

print(f'Training set size: {len(X_train)} samples')
print(f'Testing set size: {len(X_test)} samples')

print('\nSentiment distribution in training set:')
print(y_train.value_counts(normalize=True))

print('\nSentiment distribution in testing set:')
print(y_test.value_counts(normalize=True))

Training set size: 39664 samples
Testing set size: 9917 samples

Sentiment distribution in training set:
sentiment
1    0.501891
0    0.498109
Name: proportion, dtype: float64

Sentiment distribution in testing set:
sentiment
1    0.501865
0    0.498135
Name: proportion, dtype: float64


In [12]:
df = df.drop(
    columns=['review_tokens'],
    errors='ignore'
)

### Detailed Train/Test Leakage Checking

This step checks for any overlap in the unique reviews between the training and testing datasets. Even after deduplication of the original dataset, accidental leakage can occur if the split is not carefully performed, leading to an over-optimistic evaluation of the model.

In [13]:
# Convert token lists back to strings for comparison
X_train_str = X_train.apply(lambda x: ' '.join(x))
X_test_str = X_test.apply(lambda x: ' '.join(x))

# Find common reviews between train and test sets
common_reviews = set(X_train_str.unique()).intersection(set(X_test_str.unique()))
leakage_count = len(common_reviews)

print('--- Checking for Train/Test Leakage ---')
if leakage_count == 0:
    print('No leakage detected between training and testing sets.')
else:
    print(f'WARNING: {leakage_count} common reviews found between training and testing sets. This indicates potential data leakage.')
    # Optionally, print some of the leaked reviews
    # print('Leaked reviews examples:')
    # for i, review in enumerate(list(common_reviews)[:5]): # Print up to 5 examples
    #     print(f'- {review}')

--- Checking for Train/Test Leakage ---


### Post-Cleaning Duplicate Verification

Although duplicates were removed earlier, this step re-verifies that no duplicate reviews exist within the training or testing sets after all cleaning and processing. This ensures the integrity of the data used for model training and evaluation.

In [14]:
print('\n--- Post-Cleaning Duplicate Verification ---')
duplicates_in_X_train = X_train_str.duplicated().sum()
duplicates_in_X_test = X_test_str.duplicated().sum()

print(f'Number of duplicate reviews within X_train (after cleaning): {duplicates_in_X_train}')
print(f'Number of duplicate reviews within X_test (after cleaning): {duplicates_in_X_test}')

if duplicates_in_X_train == 0 and duplicates_in_X_test == 0:
    print('No duplicates found within training or testing sets after cleaning.')
else:
    print('Warning: Duplicates found within training or testing sets after cleaning. This might indicate an issue with prior deduplication steps or that the duplicates were introduced during cleaning (unlikely for this specific cleaning).')


--- Post-Cleaning Duplicate Verification ---
Number of duplicate reviews within X_train (after cleaning): 2
Number of duplicate reviews within X_test (after cleaning): 0


### TF-IDF Feature Engineering

TF-IDF (Term Frequency-Inverse Document Frequency) is a numerical statistic that is intended to reflect how important a word is to a document in a collection or corpus. It is often used as a weighting factor in information retrieval and text mining. This step transforms the text data into numerical features that can be used by machine learning models.

In [15]:
from sklearn.feature_extraction.text import TfidfVectorizer

print('\n--- TF-IDF Feature Engineering ---')
# Initialize TfidfVectorizer
# Using a reasonable max_features to control vocabulary size
tfidf_vectorizer = TfidfVectorizer(max_features=10000)

# Fit the vectorizer on the training data and transform both training and testing data
X_train_tfidf = tfidf_vectorizer.fit_transform(X_train_str)
X_test_tfidf = tfidf_vectorizer.transform(X_test_str)

print(f'Shape of X_train_tfidf: {X_train_tfidf.shape}')
print(f'Shape of X_test_tfidf: {X_test_tfidf.shape}')


--- TF-IDF Feature Engineering ---
Shape of X_train_tfidf: (39664, 10000)
Shape of X_test_tfidf: (9917, 10000)


### Saving TF-IDF Matrices

Saving the TF-IDF transformed matrices (`X_train_tfidf` and `X_test_tfidf`) allows for quick loading and reuse in subsequent modeling steps without needing to re-run the entire TF-IDF transformation process. This is particularly useful for large datasets.

In [16]:
from scipy.sparse import save_npz

print('\n--- Saving TF-IDF Matrices ---')
save_npz('X_train_tfidf.npz', X_train_tfidf)
save_npz('X_test_tfidf.npz', X_test_tfidf)
print('X_train_tfidf.npz and X_test_tfidf.npz saved.')


--- Saving TF-IDF Matrices ---
X_train_tfidf.npz and X_test_tfidf.npz saved.


### Saving the TF-IDF Vectorizer

Saving the `TfidfVectorizer` object itself is crucial. It contains the learned vocabulary and inverse document frequencies. This allows for consistent transformation of new, unseen text data (e.g., during prediction) using the same logic applied to the training data.

In [17]:
import pickle

print('\n--- Saving TF-IDF Vectorizer ---')
with open('tfidf_vectorizer.pkl', 'wb') as f:
    pickle.dump(tfidf_vectorizer, f)
print('tfidf_vectorizer.pkl saved.')


--- Saving TF-IDF Vectorizer ---
tfidf_vectorizer.pkl saved.


### Saving Feature Names

The feature names (i.e., the words corresponding to each column in the TF-IDF matrix) are useful for interpretability, allowing us to understand which words contribute most to predictions. Saving them separately makes it easy to access this information later.

In [18]:
import numpy as np

print('\n--- Saving Feature Names ---')
feature_names = tfidf_vectorizer.get_feature_names_out()
np.save('tfidf_feature_names.npy', feature_names)
print('tfidf_feature_names.npy saved.')


--- Saving Feature Names ---
tfidf_feature_names.npy saved.


### Data Quality Audit Summary

This final section provides a concise summary of the data quality checks performed, including information on potential leakage, duplicate presence, and the outcome of the feature engineering step. This serves as a quick overview of the dataset's readiness for modeling.

In [20]:
print('\n--- Data Quality Audit Summary ---')
print(f'Leakage check: {leakage_count} common reviews found.')
print(f'Duplicates within X_train: {duplicates_in_X_train}')
print(f'Duplicates within X_test: {duplicates_in_X_test}')
print(f'TF-IDF features generated: {tfidf_vectorizer.vocabulary_.__len__()}')
print('TF-IDF matrices, vectorizer, and feature names have been saved.')


--- Data Quality Audit Summary ---
Leakage check: 1 common reviews found.
Duplicates within X_train: 2
Duplicates within X_test: 0
TF-IDF features generated: 10000
TF-IDF matrices, vectorizer, and feature names have been saved.


### Removing Duplicates from Training Set

Although a general duplicate check was performed, specific duplicates within the `X_train` set were identified. This step removes those duplicates to ensure that each sample in the training data is unique, preventing potential biases in model training.

In [21]:
print('--- Removing duplicates from X_train ---')

# Identify duplicate indices in X_train_str
duplicate_indices_train = X_train_str[X_train_str.duplicated()].index

if not duplicate_indices_train.empty:
    print(f'Found {len(duplicate_indices_train)} duplicate entries in X_train. Removing them...')

    # Drop duplicates from X_train_str, X_train, and y_train
    X_train_str = X_train_str.drop(index=duplicate_indices_train)
    X_train = X_train.drop(index=duplicate_indices_train)
    y_train = y_train.drop(index=duplicate_indices_train)

    print('Duplicates removed successfully.')
    print(f'New X_train length: {len(X_train)} samples')
    print(f'New y_train length: {len(y_train)} samples')
else:
    print('No duplicates found in X_train to remove.')

--- Removing duplicates from X_train ---
Found 2 duplicate entries in X_train. Removing them...
Duplicates removed successfully.
New X_train length: 39662 samples
New y_train length: 39662 samples


In [22]:
from google.colab import files

# Save the preprocessed DataFrame to a CSV file
output_filename = 'preprocessed_imdb.csv'
df.to_csv(output_filename, index=False)

print(f'The preprocessed data has been saved to {output_filename}.')

# Provide a download link
files.download(output_filename)

The preprocessed data has been saved to preprocessed_imdb.csv.


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

In [24]:
# ============================================================
# REMOVE TRAIN / TEST OVERLAP
# ============================================================

print("=" * 60)
print("REMOVING TRAIN / TEST OVERLAP")
print("=" * 60)

train_reviews = set(X_train_str)

before_test = len(X_test)

# Find test reviews that also exist in training data
overlap_mask = X_test_str.isin(train_reviews)

overlap_count = overlap_mask.sum()

# Remove overlapping samples
X_test = X_test[~overlap_mask]
y_test = y_test[~overlap_mask]

# Keep string representation synchronized
X_test_str = X_test_str[~overlap_mask]

print("Overlapping test reviews removed:", overlap_count)

print("Test samples before:", before_test)
print("Test samples after :", len(X_test))

print("\nRemaining overlap:")

remaining_overlap = (
    set(X_train_str) &
    set(X_test_str)
)

print(len(remaining_overlap))

if len(remaining_overlap) == 0:
    print("✓ No train/test leakage remains.")
else:
    print("⚠️ Leakage still exists.")

REMOVING TRAIN / TEST OVERLAP
Overlapping test reviews removed: 1
Test samples before: 9917
Test samples after : 9916

Remaining overlap:
0
✓ No train/test leakage remains.


In [23]:
# ============================================================
# DATA QUALITY AUDIT
# ============================================================

audit = {
    'Total dataset rows': len(df),

    'Training rows': len(X_train),

    'Testing rows': len(X_test),

    'Training positive reviews':
        int((y_train == 1).sum()),

    'Training negative reviews':
        int((y_train == 0).sum()),

    'Testing positive reviews':
        int((y_test == 1).sum()),

    'Testing negative reviews':
        int((y_test == 0).sum()),

    'Duplicate reviews in dataset':
        int(df['review'].duplicated().sum()),

    'Duplicate reviews in X_train':
        int(X_train_str.duplicated().sum()),

    'Duplicate reviews in X_test':
        int(X_test_str.duplicated().sum()),

    'Train/Test overlapping reviews':
        int(len(
            set(X_train_str) &
            set(X_test_str)
        )),

    'TF-IDF features':
        int(X_train_tfidf.shape[1]),

    'Training TF-IDF rows':
        int(X_train_tfidf.shape[0]),

    'Testing TF-IDF rows':
        int(X_test_tfidf.shape[0]),

    'Training TF-IDF non-zero values':
        int(X_train_tfidf.nnz),

    'Testing TF-IDF non-zero values':
        int(X_test_tfidf.nnz),

    'TF-IDF fitted on':
        'Training data only',

    'Test transformation':
        'Training-fitted vectorizer'
}

audit_df = pd.DataFrame(
    list(audit.items()),
    columns=['metric', 'value']
)

audit_df.to_csv(
    'data_quality_audit.csv',
    index=False
)

print("✓ data_quality_audit.csv saved")

display(audit_df)

✓ data_quality_audit.csv saved


,metric,value
0,Total dataset rows,49581
1,Training rows,39662
2,Testing rows,9917
3,Training positive reviews,19906
4,Training negative reviews,19756
5,Testing positive reviews,4977
6,Testing negative reviews,4940
7,Duplicate reviews in dataset,0
8,Duplicate reviews in X_train,0
9,Duplicate reviews in X_test,0


In [26]:
# ============================================================
# REMOVE DUPLICATES FROM TRAINING SET
# ============================================================

duplicate_indices_train = X_train_str[
    X_train_str.duplicated()
].index

print(
    f"Duplicate training samples found: "
    f"{len(duplicate_indices_train)}"
)

if len(duplicate_indices_train) > 0:

    X_train = X_train.drop(
        index=duplicate_indices_train
    )

    X_train_str = X_train_str.drop(
        index=duplicate_indices_train
    )

    y_train = y_train.drop(
        index=duplicate_indices_train
    )

    print("✓ Training duplicates removed.")

else:

    print("✓ No training duplicates found.")


print("\nFinal training samples:")
print(len(X_train))

Duplicate training samples found: 0
✓ No training duplicates found.

Final training samples:
39662


In [27]:
# ============================================================
# TRAIN / TEST LEAKAGE CHECK AND REMOVAL
# ============================================================

print("=" * 60)
print("TRAIN / TEST LEAKAGE CHECK")
print("=" * 60)

train_reviews = set(X_train_str)

common_reviews = (
    train_reviews.intersection(
        set(X_test_str)
    )
)

print(
    f"Common reviews before removal: "
    f"{len(common_reviews)}"
)


# ------------------------------------------------------------
# Remove overlapping reviews from TEST set
# ------------------------------------------------------------

if len(common_reviews) > 0:

    overlap_mask = X_test_str.isin(common_reviews)

    X_test = X_test[
        ~overlap_mask
    ]

    X_test_str = X_test_str[
        ~overlap_mask
    ]

    y_test = y_test[
        ~overlap_mask
    ]

    print(
        f"✓ Removed {overlap_mask.sum()} "
        f"overlapping review(s) from test set."
    )

else:

    print("✓ No overlapping reviews found.")


# ------------------------------------------------------------
# Verify again
# ------------------------------------------------------------

remaining_overlap = (
    set(X_train_str)
    .intersection(set(X_test_str))
)

print(
    f"\nRemaining train/test overlap: "
    f"{len(remaining_overlap)}"
)

if len(remaining_overlap) == 0:
    print("✓ No train/test leakage remains.")
else:
    print("⚠️ Leakage still exists.")

TRAIN / TEST LEAKAGE CHECK
Common reviews before removal: 0
✓ No overlapping reviews found.

Remaining train/test overlap: 0
✓ No train/test leakage remains.


In [28]:
from google.colab import files

# Save the preprocessed DataFrame to a CSV file
output_filename = 'preprocessed_imdb.csv'
df.to_csv(output_filename, index=False)

print(f'The preprocessed data has been saved to {output_filename}.')

# Provide a download link
files.download(output_filename)

The preprocessed data has been saved to preprocessed_imdb.csv.


<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>